vowpalwabbit/parse_regressor.cc

```diff
@@ -54,287 +49,307 @@ const size_t buf_size = 512;
+            "\n", 1, text);
+
+        // TODO: validate skips?
+        uint32_t skip_len = (uint32_t)all.skip_strings.size();
         text_len = sprintf_s(buff, buf_size, "%d skip: ", (int)skip_len);
-		bin_text_read_write_fixed(model_file, (char *)&skip_len, sizeof(skip_len),
-				"", read, 
-			buff, text_len, text);
+        bin_text_read_write_fixed_validate_eof(model_file, (char *)&skip_len, sizeof(skip_len),
+            "", read,
+            buff, text_len, text);
 
-      for (size_t i = 0; i < skip_len; i++)
-	{
-			char skip[4] = { 0, 0, 0, 0 };
-	  if (!read) {
+        for (size_t i = 0; i < skip_len; i++)
+        {
+            char skip[4] = { 0, 0, 0, 0 };
+            if (!read) {
                 text_len = sprintf_s(buff, buf_size, "%s ", all.skip_strings[i].c_str());
-	    memcpy(skip, all.skip_strings[i].c_str(), min(3, all.skip_strings[i].size()));
-	  }
-			bin_text_read_write_fixed(model_file, skip, 3,
-				    "", read,
-				buff, text_len, text);
-	  if (read)
-	    {
-				string temp(skip);
-	      all.skip_strings.push_back(temp);
-
-				all.args.push_back("--skips");
-				all.args.push_back(boost::lexical_cast<std::string>(temp));
-	    }
-	}
-		bin_text_read_write_fixed(model_file, buff, 0,
-				"", read, 
-			"\n", 1, text);
-      
+                memcpy(skip, all.skip_strings[i].c_str(), min(3, all.skip_strings[i].size()));
+            }
+
+            bin_text_read_write_fixed_validate_eof(model_file, skip, 3,
+                "", read,
+                buff, text_len, text);
+            if (read)
+            {
+                string temp(skip);
+                all.skip_strings.push_back(temp);
+
+                all.args.push_back("--skips");
+                all.args.push_back(boost::lexical_cast<std::string>(temp));
+            }
+        }
+        bin_text_read_write_fixed_validate_eof(model_file, buff, 0,
+            "", read,
+            "\n", 1, text);
+
         text_len = sprintf_s(buff, buf_size, "options:%s\n", all.file_options->str().c_str());
-		uint32_t len = (uint32_t)all.file_options->str().length() + 1;
+        uint32_t len = (uint32_t)all.file_options->str().length() + 1;
         memcpy(buff2, all.file_options->str().c_str(), min(len, buf_size));
         
-      if (read)
+        if (read)
         {
-	len = buf_size;
+            len = buf_size;
         }
         
-		bin_text_read_write(model_file, buff2, len,
-			  "", read,
-			  buff, text_len, text);
+        bin_text_read_write_validate_eof(model_file, buff2, len,
+            "", read,
+            buff, text_len, text);
         
-      if (read)
+        if (read)
         {
-	all.file_options->str(buff2);
-    }
+            all.file_options->str(buff2);
+        }
+
+        // Read/write checksum if required by version
+        if (all.model_file_ver >= VERSION_FILE_WITH_HEADER_HASH)
+        {
+            uint32_t check_sum = uniform_hash(model_file.space.begin, model_file.space.size(), 0);
```

Comment:
This seems like it might cause a very subtle bug.  The problem is that size() is the size of the number of bytes in the buffer rather than the number of bytes in the header.  This is generally OS and FS dependent.